ruby/ql/lib/codeql/ruby/ast/Control.qll

```diff
@@ -422,6 +422,189 @@ class WhenExpr extends Expr, TWhenExpr {
   }
 }
 
+/**
+ * A `case` statement used for pattern matching.
+ */
+class CaseMatch extends ControlExpr, TCaseMatch {
```

Comment:
I was wondering whether `CaseMatch` and `CaseExpr` should be combined, but it looks like you can't actually combine `when` matching and pattern matching, e.g.
```rb
case foo
in 2
  puts "2"
when 42
  puts "42"
end
```

Still, I would expect `when` matching to be just a simple form of pattern matching.

Reply:
The Ruby interpreter actually has three different node kinds for the various flavours of `case` . The `case-in` variant raises a `NoMatchingPatternError` if none of the clauses match, while the `case-when` variants just fall through and return `nil`.